### ISSUE TITLE
Initial proposal

### ISSUE THREAD

**Issue by Azeirah** [Azeirah/nv#1]
https://github.com/nix-community/nixvim/issues/2744#issue-2759111018

Hi friendly peoples of the Nixvim world.

I have a short proposal for `nv` (pronounced "envy", ha!), a nixvim distribution aimed at Nix developers.

When you are new to development, it can get overwhelming very quickly. A new language, conventions you don't understand, what even are best practices here?

What if we had a simple and focused pre-configured distribution that is perfect for getting started with Nix development?

## Features

- [ ] Contains a high-quality formatter
- [ ] Syntax highlighting for nix and flakes
- [ ] Contains quality linting

## Batteries included

- [ ] treesitter
- [ ] telescope
- [ ] which-key
- [ ] mini.ai
- [ ] mini.surround

## Documentation

- [ ] Distributed as a flake
- [ ] Commented like [kickstart.nvim](https://github.com/nvim-lua/kickstart.nvim)

## That's it

:)

Thanks for reading

**Comment by HeitorAugustoLN** [comment]
So for the tooling, I think we should provide the following tooling by default:

### Linters (with nvim-lint)
- **deadnix**
- **statix**

### Formatters (with conform.nvim)
- **nixfmt-rfc-style**
- alejandra

### LSP
- **nixd**
- nil

###### Marked as bold is default.

**Comment by HeitorAugustoLN** [comment]
For the plugins, I think which-key is essential, for keymap discoverability, and treesitter for syntax highlighting, `mini.ai` and `mini.surround`, in my opinion, are too.

We could provide both nvim-cmp and blink.cmp for completion, but the default one I don't know yet, blink.cmp is pretty much better, but it is new, and nvim-cmp is more stable since it exists for a longer time.

**Comment by Azeirah** [comment]
`Treesitter` is indeed an absolute requirement without a doubt. So is `which-key`

I think we should focus on stability first-and-foremost.

I don't know yet about mini.ai and mini.surround, but that's mostly because I don't know much about the `mini` set of plugins yet. I would _prefer_ to go for something well known, but... yeah

I have added treesitter and which-key to the op

**Comment by Azeirah** [comment]
> So for the tooling, I think we should provide the following tooling by default:
> 
> ### Linters (with nvim-lint)
> * **deadnix**
> * **statix**
> 
> ### Formatters (with conform.nvim)
> * **nixfmt-rfc-style**
> * alejandra
> 
> ### LSP
> * **nixd**
> * nil
> 
> ###### Marked as bold is default.

We should get some matrix eyes on this list I think. I'm not experienced enough with the Nix ecosystem to be able to judge this yet.

**Comment by HeitorAugustoLN** [comment]
> I don't know yet about mini.ai and mini.surround, but that's mostly because I don't know much about the mini set of plugins yet. I would prefer to go for something well known, but... yeah

mini.nvim is pretty well known, specially for the `mini.ai`, `mini.icons`, `mini.surround` (basically vim-surround made with lua) and `mini.pairs`. kickstart.nvim uses `mini.statusline`, `mini.ai` and `mini.surround`. LazyVim uses `mini.icons`, `mini.ai`, `mini.pairs`.

From what I know, most people use at least mini.ai from mini.nvim set. So I think we should have it at least

**Comment by Azeirah** [comment]
I was aware that mini is a well-known thing, I just didn't know how _standard_ mini is, but if it is used by kickstart, then we should definitely use it 👍 

I will add it to the OP

**Comment by HeitorAugustoLN** [comment]
> We should get some matrix eyes on this list I think. I'm not experienced enough with the Nix ecosystem to be able to judge this yet.

LSP is pretty 50/50 between nixd and nil, but nixd is the more powerful from what I tested.

About the formatters, alejandra was the go-to before the nixfmt-rfc-style got usable, and nixfmt-rfc-style is the official formatter, so it should be the default.

Linters for nix there are only deadnix and statix, that both of them have different purposes, and do different linting. Statix is a more general linter, and deadnix lints if there is no dead code, mostly arguments and variables. There is no much other ways to go in linting.

**Comment by Azeirah** [comment]
Yeah, from what I recall reading in some conversations, the two linters were compatible and complementary to each other, so that sounds great :)

**Comment by Azeirah** [comment]
I'm going to make small focused issues for each "part" of the distribution

**Comment by Azeirah** [comment]
Hmmmm.. What do you think of how this should work? With kickstart the idea is that you clone the repo, right?

But since nix flakes allow you to do this however you want, we can do

- Templates
- git clone
- Use nv as an input

I think the point is that this is a starting point for people to make their own nixvim editors, just like how people do that with kickstart.

Let's say someone is very new to Nix in general, and wants their own _nv_ nixvim. What do they do?

git clone still makes the most sense, right? Because you want the "from-scratch" experience with nicely documented boilerplate.

**Comment by HeitorAugustoLN** [comment]
Templates are going to play the same role as `git clone`, and it is probably easier to set up.

**Comment by mightyiam** [comment]
I know I'd be talking about a change of scope, but why not a general purpose NixVim-based distro? That's exactly what I would like. Did you happen to listen to this episode?

https://share.transistor.fm/s/79cc70aa

**Comment by Azeirah** [comment]
> I know I'd be talking about a change of scope, but why not a general purpose NixVim-based distro? That's exactly what I would like. Did you happen to listen to this episode?
> 
> https://share.transistor.fm/s/79cc70aa

There are plenty Nixvim distros out there! This one is simply meant to decidedly _not_ be general purpose out of the box :)
